pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,531 @@
+func (rpm *RedirectPolicyManager) deleteService(svcID k8s.ServiceID) {
+	if rp, ok := rpm.services[svcID.String()]; ok {
+		// Get the policy config that selects this service.
+		config := rpm.policyConfigs[rp]
+		rpm.notifyPolicyFrontendDelete(config)
+		// Add it to the pending list in case the service is re-added.
+		rpm.pendingEvents.forService[svcID.String()] = config
+		delete(rpm.services, svcID.String())
+	}
+}
+
+// OnDeleteService handles Kubernetes service deletes, and deletes the internal state
+// for the policy config that might be associated with the service.
+func (rpm *RedirectPolicyManager) OnDeleteService(svcID k8s.ServiceID) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+
+	if rp, ok := rpm.services[svcID.String()]; ok {
+		// Get the policy config that selects this service.
+		config := rpm.policyConfigs[rp]
+		rpm.notifyPolicyFrontendDelete(config)
+		// Add it to the pending list in case the service is re-added.
+		rpm.pendingEvents.forService[svcID.String()] = config
+		delete(rpm.services, svcID.String())
+	}
+}
+
+// OnEndpointAdd handles endpoint updates, and updates internal state in case
+// the endpoint is selected by any of the current policies.
+func (rpm *RedirectPolicyManager) OnEndpointAdd(ep *endpoint.Endpoint) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+	podInfo := rpm.getEndpointPodMetadata(ep)
+	if podInfo == nil {
+		return
+	}
+	// We assume that the received events are only for node-local backends.
+	// TODO add unit test
+	// Check if any of the current redirect policies select this pod.
+	for _, config := range rpm.policyConfigs {
+		if config.policyConfigSelectsPod(podInfo) {
+			backends := rpm.getPodBackendsForPolicy(config, podInfo)
+			if len(backends) > 0 {
+				rpm.updatePolicyConfigBackends(config, true, backends...)
+			}
+		}
+	}
+}
+
+// OnEndpointAdd handles endpoint deletes, and deletes the internal state in case
+// the endpoint is selected by any of the current policies.
+func (rpm *RedirectPolicyManager) OnEndpointDelete(ep *endpoint.Endpoint) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+	podIP := ep.GetIPv4Address()
+
+	if policies, ok := rpm.podPolicy[podIP]; ok {
+		for _, podInfo := range policies {
+			config := rpm.policyConfigs[podInfo.policyID]
+			rpm.updatePolicyConfigBackends(config, false, podInfo.backends...)
+		}
+		delete(rpm.podPolicy, podIP)
+	}
+}
+
+func NewRedirectPolicyManager(svc svcManager, epMgr endpointManager) *RedirectPolicyManager {
+	return &RedirectPolicyManager{
+		frontends:     make(map[string]string),
+		services:      make(map[string]policyID),
+		podPolicy:     make(map[string][]podPolicyInfo),
+		policyConfigs: make(map[policyID]*LRPConfig),
+		svcManager:    svc,
+		epManager:     epMgr,
+		pendingEvents: eventRequests{
+			forService: make(map[string]*LRPConfig),
+		},
+	}
+}
```

Comment:
This `RedirectPolicyManager` is a great candidate for unit tests. With the passed in interfaces, it's pretty close to being able to do so.

I think there's just enough different bits flying around here that I was not able to convince myself it's trivially correct at a glance. It may be correct, it's just more code than I can quickly grok :-) . Unit tests that inject various different permutations of LRP creation / update / deletion, Service add/delete, endpoint add/delete in different orders and observing that the serviceManager is called correctly depending on the inputs would go a long way here.

Reply:
Agreed. As discussed offline, I already have the test files in progress. I would like to explore using golang/mock [1] to see if I can avoid writing too much of the boilerplate code. 
Will add unit tests in the follow-up PR. If I'm successful, we can consider using it more widely in cilium?

[1] https://github.com/golang/mock

Reply:
I think it makes sense to explore and share your experience with golang/mock , it may prove useful more broadly across the codebase. It's definitely yet another tool that developers may need to use, but if it's useful and not too hard to deal with then it may make sense to adopt.

This also might be a good topic to bring up either on Slack #develpoment channel or during cilium community meeting in open session to see if others have any experience with it and thoughts on its tradeoffs.